SourceEditorCompose/src/main/java/com/microsoft/device/display/samples/sourceeditorcompose/MainActivity.kt

```diff
@@ -0,0 +1,168 @@
+    Scaffold(
+        topBar = {
+            TopAppBar(
+                title = {
+                    if (isSinglePane) {
+                        Text(text = stringResource(id = R.string.app_name))
+                    }
+                },
+                contentColor = Color.White,
+                backgroundColor = MaterialTheme.colors.primaryVariant,
+                actions = {
+                    if (twoPaneScope.isSinglePane) {
+                        IconButton(onClick = { twoPaneScope.navigateToPane1() }) {
+                            Text(text = stringResource(R.string.editor_button))
+                        }
+                    }
+                }
+            )
+        }
+    ) {
+        AndroidView(
+            factory = {
+                WebView(it).apply {
+                    layoutParams = ViewGroup.LayoutParams(
+                        ViewGroup.LayoutParams.MATCH_PARENT,
+                        ViewGroup.LayoutParams.MATCH_PARENT
+                    )
+                    webViewClient = WebViewClient()
+                    loadData(text, "text/html", "UTF-8")
+                }
+            },
+            update = {
+                it.loadData(text, "text/html", "UTF-8")
+            }
+        )
+    }
+}
+
+/**
+ * EditorPage Composable contains the Editor for the HTML file
+ */
+@Composable
+fun TwoPaneScope.EditorPage(text: String, updateText: (String) -> Unit) {
+    val twoPaneScope = this
+    Scaffold(
+        topBar = {
+            TopAppBar(
+                title = {
+                    Text(text = stringResource(R.string.app_name))
+                },
+                backgroundColor = MaterialTheme.colors.primaryVariant,
+                contentColor = Color.White,
+                actions = {
+                    if (twoPaneScope.isSinglePane) {
+                        IconButton(onClick = { twoPaneScope.navigateToPane2() }) {
+                            Text(text = stringResource(id = R.string.preview_button))
+                        }
+                    }
+                }
+            )
+        }
+    ) {
+        TextField(
+            value = text,
+            onValueChange = { newText -> updateText(newText) },
+            modifier = Modifier.fillMaxSize()
+        )
+    }
+}
+
+/**
+ * For handling window navigation for Surface Duo, this acts as a main Controller for both the Composable
+ */
+@Composable
+fun ControllerLayer() {
+    val context = LocalContext.current
+    val fileName = "Source.html"
+    var text by rememberSaveable { mutableStateOf(readFile(fileName, context)) }
+    val updateText: (String) -> Unit = { newText -> text = newText }
+
+    TwoPaneLayout(
+        pane1 = { EditorPage(text = text, updateText = updateText) },
+        pane2 = { PreviewPage(text = text) }
+    )
+}
+
+/**
+ * To read a file from assets folder, which currently contains the local Source.html file
+ */
+private fun readFile(file: String, context: Context?): String {
```

Comment:
AndroidStudio is suggesting that the filename value is always "Source.html" - if you like, you can just remove the parameter and set fileName internally inside this method